### PR TITLE
HIVE-23891: Using UNION sql clause and speculative execution can cause file duplication in Tez

### PR DESCRIPTION
**Pull request by georgepachitariu**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 1]
```diff
@@ -1459,8 +1459,12 @@ public static void mvFileToFinalPath(Path specPath, Configuration hconf,
       }
 
       // Remove duplicates from tmpPath
-      List<FileStatus> statusList = HiveStatsUtils.getFileStatusRecurse(
-          tmpPath, ((dpCtx == null) ? 1 : dpCtx.getNumDPCols()), fs);
+      int level = dpCtx == null ? 1 : dpCtx.getNumDPCols();
+      // - when execution engine is Tez and the query uses sql clause "Union"
+      //   there is an extra layer of subdirectories that contains the branches of the union;
+      boolean isUnionClauseInTez = conf.getDirName().toString().contains(AbstractFileMergeOperator.UNION_SUDBIR_PREFIX);
```
Comment (pgaref) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1465]: Good catch!

File: ql/src/test/queries/clientpositive/insert_with_union_and_tez.q [hunk 2]
```diff
@@ -0,0 +1,16 @@
+set tez.am.speculation.enabled=true;
```
Comment (pgaref) [inline ql/src/test/queries/clientpositive/insert_with_union_and_tez.q:1]: How is AM speculative execution affecting this behaviour?

Comment (georgepachitariu) [reply ql/src/test/queries/clientpositive/insert_with_union_and_tez.q:1]: Oh sorry! I thought it's for the tasks, I will remove it and push again.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 1 above)
Comment (georgepachitariu) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1465]: Thank you :)

**Comment by georgepachitariu** [comment]
Hello @kgyrtkirk,
Can you please guide me on what I should do next?

I saw that there is a new test which is timing-out in `continuous-integration/jenkins/pr-merge`, but I think that is unrelated to my change.
Thank you, George.

**Comment by georgepachitariu** [comment]
Hello @kgyrtkirk,
Can you please have a look?

**Comment by dengzhhu653** [comment]
Hello @georgepachitariu, cloud you please submit another PR against the latest master? Thank you!

**Comment by abstractdog** [comment]
handled in https://github.com/apache/hive/pull/3836

**Pull request closed without merging**